### PR TITLE
fix: fixed marker flickering caused by creating new instances on each render

### LINKED ISSUE #439: Markers and Clusters are redraw when zoom is changing
I reviewed this example `http://jsfiddle.net/geocodezip/f5baj1uq/`
it uses `https://cdnjs.cloudflare.com/ajax/libs/markerclustererplus/2.1.4/markerclusterer.js`
I love it because I zoom markers and clusters and them are not redrawn.
[MarkerCluster-Ok.webm](https://user-images.githubusercontent.com/67704982/192104052-b6b67bc5-cc0b-44a8-bed3-1b862caaf7fe.webm)

But checking the last version `https://googlemaps.github.io/js-markerclusterer/public/defaults/` I see that all markers and clusters are drawn for example in each zoom changing giving a bad user experience because unnecessary markers and clusters are drawing again.
[MarkerCluster-Ko.webm](https://user-images.githubusercontent.com/67704982/192104133-b37be700-7e72-44f8-898c-b9b8bb91f1cb.webm)

Is there an option in the last version to avoid flash/redraw effect in markers and clusters? 
Thanks for your response.

### PR DESCRIPTION
**Pull request by nora-soderlund**
This PR aims to resolve #439 by reusing the previous markers where possible, by comparing exact coordinates on clusters with a "processed" marker (Cluster.marker).

This also fixes part of #384. This will keep the bounce animation while zooming in and out, but it will stop when the marker is turned into a cluster and won't bounce when turned back into a marker again, [explained here](https://github.com/googlemaps/js-markerclusterer/issues/384#issuecomment-1210036844).

https://user-images.githubusercontent.com/78360666/213556345-4c9ac6b1-e6dc-452e-bd19-7ee5e3ae08db.mp4

This is how it was previously, taken from #439:

> [192104133-b37be700-7e72-44f8-898c-b9b8bb91f1cb.webm](https://user-images.githubusercontent.com/78360666/213556817-2e0e70e3-065e-4ccf-963f-a74e9c469895.webm)

### CODE REVIEW

**Comment by daun** [comment]
Can confirm this helps a lot to reduce flicker on zoom! Would love to see this merged or as an optional setting.

**Comment by amuramoto** [comment]
@nora-soderlund apologies for the slow reply - this looks great. Can you add a test to cover it then I can merge?

**Comment by nora-soderlund** [comment]
Will do next week.

**Comment by vicb** [comment]
I am seeing flicker with advanced markers (on v2.1.3).

Would fixing this issue also impacts advanced markers?

Let's try to move this forward!

File: src/markerclusterer.ts [hunk 1]
```diff
@@ -178,8 +178,22 @@ export class MarkerClusterer extends OverlayViewSafe {
 
       // allow algorithms to return flag on whether the clusters/markers have changed
       if (changed || changed == undefined) {
+        clusters.forEach((cluster, index) => {
+          // if there's clusters that has been "processed" (Cluster.marker is set), then reuse it instead of reprocessing it
+          const duplicate = this.clusters.findIndex((old) => old.marker && old.marker.getPosition().equals(cluster.position));
```
Comment (vicb) [inline src/markerclusterer.ts:183]: It looks like the only change to support advanced markers would be to use `MarkerUtils.getPosition()` here

https://github.com/googlemaps/js-markerclusterer/blob/9476b3a62399fd124c92ed74faae6943cf573ea4/src/marker-utils.ts#L41-L60

File: src/markerclusterer.ts [hunk 2]
```diff
@@ -222,6 +236,10 @@ export class MarkerClusterer extends OverlayViewSafe {
       if (cluster.markers.length === 1) {
         cluster.marker = cluster.markers[0];
       } else {
+        // we disable the markers here only when we're re-processing the cluster marker
+        // this stops the markers from flickering on each map event.
+        cluster.markers.forEach((marker) => marker.setMap(null));
```
Comment (vicb) [inline src/markerclusterer.ts:241]: OOps this one would needs to be update to to use

https://github.com/googlemaps/js-markerclusterer/blob/9476b3a62399fd124c92ed74faae6943cf573ea4/src/marker-utils.ts#LL33C1-L39C4

File: src/markerclusterer.ts (hunk 1 above)
Comment (vicb) [inline src/markerclusterer.ts:183]: There might be some possible optimization here.

We are searching the former clusters for each new cluster. So basically o(n^2).

Using a map of positionString -> cluster should speed that up.

**Comment by nora-soderlund** [comment]
> Will do next week.

image

_oops_, looking into this now!

File: src/markerclusterer.ts [hunk 3]
```diff
@@ -180,8 +180,23 @@ export class MarkerClusterer extends OverlayViewSafe {
 
       // allow algorithms to return flag on whether the clusters/markers have changed
       if (changed || changed == undefined) {
+        clusters.forEach((cluster, index) => {
+          // if there's clusters that has been "processed" (Cluster.marker is set), then reuse it instead of reprocessing it
+          const duplicate = this.clusters.findIndex((old) => old.marker && MarkerUtils.getPosition(old.marker).equals(cluster.position));
```
[the remark this answers is not in the document]
Comment (nora-soderlund) [reply src/markerclusterer.ts:185]: Yep, iirc I was experimenting with indices before coming to a solution, probably why it was left as an index.

/Updated!

**Comment by vicb** [comment]
Thanks for all the updates!

I think it might be worth trying the map solution I proposed in a comment, something like:

```js
function clusterKey(cluster: Cluster) {
  const position = cluster.getPosition();
  return `${position.lat()}#${position.lng()}`;
}
```

```js
// ...
        const clusterMap = new Map<string, Marker>();
        this.clusters.forEach(cluster => {
          if (cluster.marker) {
            clusterMap.set(MarkerUtils.clusterKey(cluster), cluster);
          }        
        });  
    
        clusters.forEach((cluster, index) => {
          const clusterKey = clusterKey(cluster);          
          const duplicateCluster = clusterMap.get(clusterKey);

          if(duplicateCluster) {
            // we'll reset the Cluster.markers here and only add the existing item
            // down below in renderClusters, it will assign this one item to Cluster.marker
            clusters[index].markers.length = 0;
            clusters[index].markers.push(duplicateCluster.marker);
            duplicateCluster.marker = null;
            clusterMap.delete(clusterKey);
          }
        });
// ...
```

At least time it vs `find` with a large nb of markers and see what's more efficient

**Comment by nora-soderlund** [comment]
> Thanks for all the updates!
> 
> I think it might be worth trying the map solution I proposed in a comment, something like:
> 
> // ...
> 
> At least time it vs `find` with a large nb of markers and see what's more efficient

At the very least **in my browser**, string comparisons took up to 27 times longer:
```
iterating 1000 searches for string keys
search took 4565ms
iterating 1000 searches for latLng objects
search took 165ms
```

This is what I tested:
```js
    const keys = [];
    const latLngs = [];

    for(let index = 0; index < 10000; index++) {
      const lat = Math.random();
      const lng = Math.random();

      const latLng = new google.maps.LatLng({
        lat,
        lng
      });

      keys.push(`${lat}x${lng}`);
      latLngs.push(latLng);
    }

    console.log("iterating 1000 searches for string keys");

    let start = performance.now();

    for(let index = 0; index < 1000; index++) {
      keys.find((key) => key === `${(Math.random())}x${(Math.random())}`);
    }

    let end = performance.now();

    console.log(`search took ${(end - start)}ms`);

    console.log("iterating 1000 searches for latLng objects");

    start = performance.now();

    for(let index = 0; index < 1000; index++) {
      const random = new google.maps.LatLng({ lat: Math.random(), lng: Math.random() });

      latLngs.find((latLng) => random.equals(latLng));
    }

    end = performance.now();

    console.log(`search took ${(end - start)}ms`);
```

**Comment by nora-soderlund** [comment]
@amuramoto ready for review, added advanced marker support and a proper test case.

**Comment by vicb** [comment]
@nora-soderlund thanks for trying. That was my fear but we can do better, switching from:

```
const clusterMap = new Map<string, Marker>();
```

to

```
const clusterMap = new Map<number, Map<number, Marker>>();
```

The idea would be lat -> lng -> Marker.

It should be both way faster and more memory efficient.

**Comment by vicb** [comment]
Also I am not sure what you tests, can't see a `Map` there?

**Comment by amuramoto** [comment]
@vicb Since this PR resolves the flickering issue as described, how about we merge and open a new issue on the potential optimizations and continue this there?

**Comment by vicb** [comment]
@amuramoto I saw a clear during my tests - o(n^2) can become slow with big ns.

**Comment by nora-soderlund** [comment]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map works differently than I thought, it is indeed quicker. Updated.

**Comment by vicb** [comment]
I think https://github.com/googlemaps/js-markerclusterer/pull/536#issuecomment-1561733953 is worth considering.

It will probably be a few more LOCs but should be faster (do not create string keys, comparing number vs long string) and more memory efficient (numbers take way less memory than this string key).

**Comment by vicb** [comment]
And also you need to `record.delete()` to only re-use a cluster once - see my original snippet

**Comment by nora-soderlund** [comment]
I will leave it as it is now.

File: src/markerclusterer.ts [hunk 4]
```diff
@@ -180,8 +180,40 @@ export class MarkerClusterer extends OverlayViewSafe {
 
       // allow algorithms to return flag on whether the clusters/markers have changed
       if (changed || changed == undefined) {
+        const records = new Map<{ lat: number; lng: number }, Cluster>();
```
Comment (vicb) [inline src/markerclusterer.ts]: image

That does not work in JS.

File: src/markerclusterer.ts [hunk 5]
```diff
@@ -180,8 +180,40 @@ export class MarkerClusterer extends OverlayViewSafe {
...
+
+        this.clusters.forEach(
+          (cluster) =>
+            cluster.marker &&
```
Comment (vicb) [inline src/markerclusterer.ts:187]: not sure if there is any coding standards in this repo (@amuramoto ?) but personally prefer the longer `if () {...}` version

File: src/markerclusterer.ts (hunk 4 above)
Comment (nora-soderlund) [reply src/markerclusterer.ts]: [b0bf346](https://github.com/googlemaps/js-markerclusterer/pull/536/commits/b0bf346064a9a209cc3135d7eb8b36b369ad7574)

File: src/markerclusterer.ts (hunk 5 above)
Comment (nora-soderlund) [reply src/markerclusterer.ts:187]: eslinted.

**Comment by vicb** [comment]
@nora-soderlund I can take over to implement `new Map<number, Map<number, Marker>>()` if you prefer. Will need to find some time over the next couple days. Let me know.

**Comment by vicb** [comment]
I had some time to give this CL more thoughts.

**What the original code does:**

In `MarkerClusterer.render` when the `Cluster`s have changed they are all removed from the map at once (via `this.reset()`) and then re-added to the map one by one in `renderClusters` via `MarkerUtils.setMap(cluster.marker, map)`.

The flicker is caused by removing + re-adding the `Cluster`s.

Note that a `Cluster` (i.e. the class name) could either be an individual maker or a "Cluster" (=group) of marker represented by a a marker created by the renderer.

**What this CL changes:**

This CL do no remove all the markers from the map.

It would only remove a marker is there was no marker at the exact same (lat, lon) in thr previous render iteration.

This is fine when the `Cluster` contains a single marker.
However it has corner cases when the `Cluster` was a group of multiple marker.
I added one example in my previous message: the `Cluster` can be at the same position but have a different number of markers.

But more generally a group icon depends on more things: `this.renderer.render(cluster, stats, map)`. `stats` here should also be taken into account. Let's imagine you are zooming in or out but the grouping does not change. It is still likely that the `stats` would change.

In conclusion I think that the assumption of a group position not changing meaning that the rendering does not change is flawed. The count can change, the `stats` can change.

I think this can be solved by adding a `Cluster.shouldRerender(stats)` method. This could be an optional method to keep backward compatibility. If the method is not implemented by the renderer we should keep a conservative approach and re-render groups every time.

Thoughts?

**Comment by amuramoto** [comment]
@vicb  I think there may be an easy solution here that relates to your comment 'This is fine when the `Cluster` contains a single marker.' Assuming two things:

1. The goal here is to get the individual markers to not constantly re-render which causes flickering
2. Actual 'clusters', i.e. clusters where `Cluster.markers > 1` likely should always re-render anyway since their position is influenced by zoom-level change, and/or the addition/subtraction when the number of markers in the cluster changes

So, can't we just adjust what's in this PR to only do the `duplicateCluster` check if `Cluster.markers === 1`? This would apply the logic to only individual markers and not clusters of multiple markers, which I think avoids the edge case you're describing.

On a side note, I think there is one additional edge case in this PR. Since `records` is a `Map`, we can't track duplication of multiple markers that are at the same lat/lng. Should be fairly easy to solve that tho.

**Comment by amuramoto** [comment]
Sounds like a plan to me!

@nora-soderlund Any thoughts on this approach? Me and @vicb would be happy to contribute the changes to this PR

**Comment by nora-soderlund** [comment]
Feel free to push changes. :)

**Comment by vicb** [comment]
It took some discussion but I really like the solution we are shaping together.

Starting with `Cluster`s with only one marker was a really great idea.

I think the implementation could be very simple.
There is no need to check the marker position.
The only question we need to answer is: was this marker already rendered?

When it was then there is nothing to do. That is no need to remove/add it back (via `setMap(null)`/ `setMap(map)`) which means no flicker.

When it was not, we only need to add it to the map (via `setMap(map)`).

Checking the position would still be useful for cluster with > 1 markers.

I probably won't have time to work on that until next week.

**Comment by amuramoto** [comment]
Thank you for your work on this @nora-soderlund !
